src/awkward/_v2/contents/listarray.py

```diff
@@ -151,14 +151,136 @@ def _carry(self, carry, allow_lazy, exception):
             self._parameters,
         )
 
+    def compact_offsets64(self, start_at_zero):
+        nplike = self.nplike
+        out = ak._v2.index.Index64.empty(len(self._starts) + 1, self.nplike)
+        self._handle_error(
+            nplike[
+                "awkward_ListArray_compact_offsets",
+                out.dtype.type,
+                self._starts.dtype.type,
+                self._stops.dtype.type,
+            ](
+                out.to(nplike),
+                self._starts.to(nplike),
+                self._stops.to(nplike),
+                len(self._starts),
+            )
+        )
+        return out
+
+    def broadcast_tooffsets64(self, offsets):
+        nplike = self.nplike
+        if len(offsets) == 0 or offsets[0] != 0:
+            raise ValueError(
+                "broadcast_tooffsets64 can only be used with offsets that start at 0"
+            )
+
+        if len(offsets) - 1 != len(self._starts):
+            raise ValueError(
+                "cannot broadcast ListArray of length ",
+                self._length,
+                " to length ",
+                len(offsets) - 1,
+            )
+
+        carrylen = offsets[-1]
+        nextcarry = ak._v2.index.Index64.empty(carrylen, nplike)
+
+        self._handle_error(
+            nplike[
+                "awkward_ListArray_broadcast_tooffsets",
+                nextcarry.dtype.type,
+                offsets.dtype.type,
+                self._starts.dtype.type,
+                self._stops.dtype.type,
+            ](
+                nextcarry.to(nplike),
+                offsets.to(nplike),
+                len(offsets),
+                self._starts.to(nplike),
+                self._stops.to(nplike),
+                len(self._content),
+            )
+        )
+        nextcontent = self._content._carry(nextcarry, True, NestedIndexError)
+        if self._identifier is not None:
+            identifier = self._identifier[slice(0, len(offsets) - 1)]
+        else:
+            identifier = self._identifier
+        return ak._v2.contents.listoffsetarray.ListOffsetArray(
+            offsets, nextcontent, identifier, self._parameters
+        )
+
+    def toListOffsetArray64(self, start_at_zero):
+        offsets = self.compact_offsets64(start_at_zero)
+        out = self.broadcast_tooffsets64(offsets)
+        return out
+
+    def _getitem_next_jagged(self, slicestarts, slicestops, slicecontent, tail):
```

Comment:
It's a better idea to not introduce `_getitem_next_jagged` until this PR, adding old tests and their consequences, is done. You don't want a single PR to have to deal with both updates.

None of the tests in the list should involve jagged slicing. If one does, then I've misidentified it. (And since jagged slicing was introduced in #111, it's hard to see how any test with a number smaller than that could include it.)